ibis/pandas/execution.py

```diff
@@ -0,0 +1,224 @@
+import numbers
+import operator
+
+from functools import reduce
+
+import six
+
+import toolz
+
+import pandas as pd
+
+import ibis.expr.types as ir
+import ibis.expr.operations as ops
+
+from ibis.pandas.dispatch import execute, execute_node
+
+
+@execute_node.register(ir.Literal)
+def execute_node_literal(op):
+    return op.value
+
+
+@execute_node.register(ir.Literal, object)
+def execute_node_literal_object(op, _, scope=None):
+    return op.value
+
+
+@execute.register(ir.Literal)
+def execute_literal(literal):
+    return literal.value
+
+
+@execute_node.register(ops.TableColumn, pd.DataFrame)
+def execute_table_column_dataframe(op, data, scope=None):
+    return data[op.name]
+
+
+@execute_node.register(ops.Selection, pd.DataFrame)
+def execute_selection_dataframe(op, data, scope=None):
+    selections = op.selections
+    predicates = op.predicates
+    sort_keys = op.sort_keys
+
+    result = data
+
+    if selections:
+        old_names = [s.op().name for s in selections]
+        new_names = [
+            getattr(s, '_name', old_name) or old_name
+            for old_name, s in zip(old_names, selections)
+        ]
+        result = result[old_names].rename(dict(zip(old_names, new_names)))
+
+    if predicates:
+        where = reduce(operator.and_, (execute(p, scope) for p in predicates))
+        result = result.loc[where]
+
+    if sort_keys:
+        return result.sort_values(sort_keys)
+    return result
+
+
+def reduction_to_lambda(expr):
+    op = expr.op()
+
+    def reduction(
+        df,
+        column=op.args[0]._name,
+        method_name=type(op).__name__.lower(),
+        new_name=expr._name,
+    ):
+        method = getattr(df[column], method_name)
+        return method().rename(new_name)
+    return reduction
+
+
+@execute_node.register(ops.Aggregation, pd.DataFrame)
+def execute_aggregation_dataframe(op, data, scope=None):
+    assert op.agg_exprs
+    assert op.by
+
+    predicates = op.predicates
+    if predicates:
+        data = data.loc[
+            reduce(operator.and_, (execute(p, scope) for p in predicates))
+        ]
+
+    gb = data.groupby([execute(by, scope) for by in op.by])
+
+    aggs = [reduction_to_lambda(agg_expr) for agg_expr in op.agg_exprs]
+    pieces = [func(gb) for func in aggs]
+    agg_df = pd.concat(pieces, axis=1)
+
+    having = op.having
+    if having:
+        agg_df = agg_df[
+            reduce(operator.and_, (execute(h, scope) for h in having))
```

Comment:
This needs a test and I suspect this won't work as is because there's no data being bound to the parent expression. This should be solved by passing in scope mapping the parent of the having expr to `agg_df`.